narwhal/consensus/src/bullshark.rs

```diff
@@ -155,21 +156,90 @@ impl Bullshark {
             return Ok((Outcome::LeaderBelowCommitRound, Vec::new()));
         }
 
+        let mut committed_sub_dags = Vec::new();
+        let outcome = loop {
+            let (outcome, committed) = self.commit_leader(leader_round, state)?;
+
+            // always extend the returned sub dags
+            committed_sub_dags.extend(committed);
+
+            // break the loop and return the result as long as there is no schedule change.
+            // We want to retry if there is a schedule change.
+            if outcome != Outcome::ScheduleChanged {
+                break outcome;
```

Comment:
This is a nitpick and I doubt it need fixing but wanted to raise it to agree on the approach:

The current code when it detects a leader schedule it correctly aborts all pending leaders after the schedule. Then it goes back to the last round we have in the DAG and tries to see if it can commit that leader.  

If it succeeds than we are good and we commit any intermediate leaders we could commit. But if that specific round has not leader it exits.
This means that even if we could commit leaders between **leader_schedule_change_round** and **leader_round** we do not check it which might have an impact in the latency. We can keep it that way or do a backpass on all rounds to see if we can find any intermediate leaders to commit

Reply:
Very good point, I can definitely look into this optimization. My instinct is that it will increase complexity (and I primarily went for simplicity at least for this iteration), but if things are still manageable then it worth covering that case.

Reply:
Keeping things simple is fine for now. FWIW, we just need to find the highest round leader (between leader_schedule_change_round and leader_round) under the new schedule that can commit.